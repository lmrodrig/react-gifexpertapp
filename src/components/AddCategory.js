import React, { useState } from 'react'
import PropTypes from 'prop-types';   
 

export const AddCategory = ({setCategories}) => {
    //el estado default se debe de declarar sino sera undefined
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) =>{
        setInputValue(e.target.value);
        //console.log('HandleInput changed...!');

    }

    const handleSubmit = (e) =>{
        //para quedarnos en la misma ruta sin agregar /? al path
        e.preventDefault();
        if(inputValue.trim().length > 2){
            setCategories( cats => [inputValue,...cats]);
            setInputValue('');
        }
        //console.log('Se llamo el handleSubmit..');
        
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
           <input
                type='text'
                value={inputValue}
                onChange={handleInputChange}
           />
        </form>
    )
    
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired 
    
}