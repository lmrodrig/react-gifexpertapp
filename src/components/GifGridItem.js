import React from 'react'

export const GifGridItem = ( { title, url}) => {

    //react toma className como class de CSS
    return (
        <div className='card animate__animated animate__fadeIn'>
            <img src={url} alt={title}></img>
            <p>{title}</p>
        </div>
    )
}
