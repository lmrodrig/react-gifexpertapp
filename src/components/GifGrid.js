import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types';  

import { useFetchGifs } from "../hooks/UseFetchGifs";
import { GifGridItem } from './GifGridItem';
//import { getGifs } from '../helpers/GetGifs';

export const GifGrid = ({ category}) => {

    
    //const [images, setImages] = useState([]);
    //Este Hook dice que solo se debe de ejecutar 1 sola vez
    //cuando se inicializa el componente.
    // useEffect(() => {
    //     //haremos una peticion HTTP con fetch
    //     getGifs(category)
    //     .then( imgs => setImages(imgs));
    // }, [category]); // Aqui solo le indica que si la categoria cambia se lanza otra vez
    //                 // la llamada a getGifs 
        
    // en el uso del map se pide tener una key unica en el GifGridItem
    // por eso se pone como propiedad del elemento
    // se puede confundir que es un parametro, pero no lo es. 

    //usando custom hooks
    const { data:images, loading } = useFetchGifs( category );

    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> { category } </h3>

            { loading && <p className="animate__animated animate__flash">Loading</p> }

            <div className="card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}