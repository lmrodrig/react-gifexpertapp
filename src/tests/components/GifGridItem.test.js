import React from "react";
//import '@testing-library/jest-dom'
import { shallow } from 'enzyme'

import {GifGridItem} from '../../components/GifGridItem'

//Para usar el enzyme
/* Referencias:
        https://enzymejs.github.io/enzyme/
        https://github.com/wojtekmaj/enzyme-adapter-react-17 -> este quiza no es necesario
        https://www.npmjs.com/package/enzyme-to-json 
        
        */
        /* Se debe de instalar
            *npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 -> este quiza no es necesario
            npm i --save-dev enzyme enzyme-adapter-react-16
            npm install --save-dev enzyme-to-json
        */
        /* se debe de configurar en setupTests.js
            import Enzyme from 'enzyme';
            import Adapter from 'enzyme-adapter-react-16';
            import {createSerializer} from 'enzyme-to-json';

            Enzyme.configure({ adapter: new Adapter() });

            expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));
        
        */

describe('Pruebas en <GifGridItem /> component', () => {
    const titulo = 'Titulo';
    const url = 'https://localhost/algo.jpg';
    const wrapper = shallow( <GifGridItem title={titulo} url={url} />);
        
    test('Debe de mostrar < GifGridItem /> correctamente', () => {
        
        //se crea un snapshot en la carpeta de test
        expect (wrapper).toMatchSnapshot();
    })

    test('Debe de tener un Parrafo con el titulo', () => {
      
        const p= wrapper.find('p');
        expect ( p.text()).toBe(titulo);

    });

    test('Debe de tener la imagen url y alt igual a los props', () => {
      const img = wrapper.find('img');
      //console.log ( img.html());  // es para logear tal cual el html
      //console.log ( img.props()); // Para ver las propiedades de ese elemento  
      expect ( img.prop('src')).toBe(url);
      expect ( img.prop('alt')).toBe(titulo);

    });
    
    test('Debe de contener la clase animate__fadeIn', () => {
      
      const class1= 'animate__fadeIn';  
      const div = wrapper.find('div');
       
      //console.log ( div.prop('className'));
      expect( div.prop('className')).toContain(class1);

      //misma validacion pero usando boolean
      const className = div.prop('className');
      expect (className.includes(class1)).toBe(true);

    });
    
    
});
