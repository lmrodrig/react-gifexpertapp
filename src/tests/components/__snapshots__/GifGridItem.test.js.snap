// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem /> component Debe de mostrar < GifGridItem /> correctamente 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="Titulo"
    src="https://localhost/algo.jpg"
  />
  <p>
    Titulo
  </p>
</div>
`;
